### PR TITLE
[ETHEREUM-CONTRACTS] upgradeTo logic "fix" - really a slight change

### LINKED ISSUE #1201: [ETHEREUM-CONTRACTS] Amend SuperToken._upgrade logic
## Technical notes

- "if `to` is diffferent from `account`, we requireReceptionAck"
  - we think it's good idea to eliminate it.
  - maybe we should also take into account if "data" is empty or not.

### PR DESCRIPTION
**Pull request by 0xdavinchee**
- pass `userData` to `_upgrade`'s `userData` field, not `operatorData` field
- new logic which checks if `userData.length > 0` instead of `account != to` for `requireReceptionAck` parameter passed to `_mint` => `_callTokensReceived`

Tests  added/modified to check new logic:
- tests added to ensure this doesn't block EOAs despite userData
- tests added to ensure this doesn't block non-erc777-registered contracts w/ no userData passed
- tests added to ensure this blocks non-erc777-registered contracts w/ userData passed

cc @d10r @ngmachado

### CODE REVIEW

File: packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperToken.sol [hunk 1]
```diff
@@ -388,6 +388,9 @@ interface ISuperToken is ISuperfluidToken, TokenInfo, IERC20, IERC777 {
      *
      * @custom:note It will use `transferFrom` to get tokens. Before calling this
      * function you should `approve` this contract
+     * 
+     * @custom:requirements
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperToken.sol:394]: :+1:

**Review by hellwolf — COMMENTED** [review]
looks good, @d10r  please ack and be aware too.

File: packages/ethereum-contracts/contracts/superfluid/SuperToken.sol [hunk 2]
```diff
@@ -601,8 +601,8 @@ contract SuperToken is
         if (underlyingAmount != actualUpgradedAmount) revert SUPER_TOKEN_INFLATIONARY_DEFLATIONARY_NOT_SUPPORTED();
 
         _mint(operator, to, adjustedAmount,
-            // if `to` is diffferent from `account`, we requireReceptionAck
-            account != to, userData, operatorData);
+            // if `userData.length` than 0, we requireReceptionAck
+            userData.length > 0, userData, operatorData);
```
Comment (ngmachado) [inline packages/ethereum-contracts/contracts/superfluid/SuperToken.sol]: to save gas lets use `!=`

**Review by ngmachado — DISMISSED** [review]

File: packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperToken.sol (hunk 1 above)
Comment (hellwolf) [reply packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperToken.sol:394]: add a @custom:warning about the potential re-entrance case if the "to" account is ERC-777.

File: packages/ethereum-contracts/contracts/superfluid/SuperToken.sol (hunk 2 above)
Comment (0xdavinchee) [reply packages/ethereum-contracts/contracts/superfluid/SuperToken.sol]: resolved mr

File: packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperToken.sol (hunk 1 above)
Comment (0xdavinchee) [reply packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperToken.sol:394]: resolved mr

**Review by hellwolf — APPROVED** [review]

**Pull request merged**